libraries/AP_RSSI/AP_RSSI.cpp

```diff
@@ -203,44 +205,81 @@ float AP_RSSI::read_channel_rssi()
     return channel_rssi;    
 }
 
-// read the PWM value from a pin
-float AP_RSSI::read_pwm_pin_rssi()
+void AP_RSSI::check_pwm_pin_rssi()
 {
-#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
-    // check if pin has changed and initialise gpio event callback
-    pwm_state.gpio = get_gpio(rssi_analog_pin);
-    if (pwm_state.gpio != pwm_state.last_gpio) {
-
-        // remove old gpio event callback if present
-        if (pwm_state.last_gpio != 0) {
-            stm32_gpiosetevent(pwm_state.last_gpio, false, false, false, nullptr);
-            pwm_state.last_gpio = 0;
-        }
+    if (rssi_analog_pin == pwm_state.last_rssi_analog_pin) {
+        return;
+    }
 
-        // install interrupt handler on rising or falling edge of gpio
-        if (pwm_state.gpio != 0) {
-            stm32_gpiosetevent(pwm_state.gpio, true, true, false, irq_handler);
-            pwm_state.last_gpio = pwm_state.gpio;
+    // detach last one
+    if (pwm_state.last_rssi_analog_pin) {
+        if (pwm_state.last_rssi_analog_pin &&
+            !hal.gpio->attach_interrupt(pwm_state.last_rssi_analog_pin,
+                                        nullptr,
+                                        AP_HAL::GPIO::INTERRUPT_NONE)) {
+            gcs().send_text(MAV_SEVERITY_WARNING,
+                            "RSSI: Failed to detach from pin %u",
+                            pwm_state.last_rssi_analog_pin);
+            // ignore this failure or the user may be stuck
         }
     }
 
-    // disable interrupts temporarily
-    irqstate_t istate = irqsave();
+    pwm_state.last_rssi_analog_pin = rssi_analog_pin;
 
-    // check for timeout
-    float ret;
-    if ((pwm_state.last_reading_ms == 0) || (AP_HAL::millis() - pwm_state.last_reading_ms > 1000)) {
-        pwm_state.value = 0;
-        ret = 0;
-    } else {
-        // convert pwm value to rssi value
-        ret = scale_and_constrain_float_rssi(pwm_state.value, rssi_channel_low_pwm_value, rssi_channel_high_pwm_value);
+    if (!rssi_analog_pin) {
+        // don't need to install handler
+        return;
+    }
+
+    // install interrupt handler on rising and falling edge
+    if (!hal.gpio->attach_interrupt(
+            rssi_analog_pin,
+            FUNCTOR_BIND_MEMBER(&AP_RSSI::irq_handler,
+                                void,
+                                uint8_t,
+                                bool,
+                                uint32_t),
+            AP_HAL::GPIO::INTERRUPT_BOTH)) {
+        // failed to attach interrupt
+        gcs().send_text(MAV_SEVERITY_WARNING,
+                        "RSSI: Failed to attach to pin %u",
+                        rssi_analog_pin);
+        return;
     }
+}
+
+// read the PWM value from a pin
+float AP_RSSI::read_pwm_pin_rssi()
+{
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN || CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
```

Comment:
Again, I don't see a good reason to put board checks here, it's using a HAL method.